File: heapsort.py
"""
Heapsort algorithm implemented in Python 3.
"""

class HeapSort:
    def __init__(self, array):
        self.array = array
        self.length = len(array)

    def swap(self, i, j):
        tmp = self.array[i]
        self.array[i] = self.array[j]
        self.array[j] = tmp

    def adjust_heap(self, i, n):
        while 2*i + 1 < n:
            child = 2*i + 1
            
            if child + 1 < n and self.array[child + 1] > self.array[child]:
                child = child + 1

            if self.array[i] >= self.array[child]:
                return
            
            self.swap(i, child)
            i = child
        
    def construct_heap(self):
        i = (self.length - 1) // 2
        while i >= 0:
            self.adjust_heap(i, self.length)
            i = i - 1

    def sort(self):
        self.construct_heap()
        for i in range(0, self.length):
            self.swap(0, self.length-i-1)
            self.adjust_heap(0, self.length-i-1)


    def dump(self):
        for i in range(0, self.length):
            print(self.array[i], end=" ")
        print("")

File: src/cpp/heapsort.cc
#include<vector>
#include<iostream>
#include "heapsort.h"


using namespace std;

void swap(vector<int> &v, int i, int j)
{
  int tmp = v[i];
  v[i] = v[j];
  v[j] = tmp;
}

void heapify(vector<int> &v, int i, int n)
{
  while (2*i + 1 < n)
  {
    int left = 2*i + 1;
    if ((left + 1 < n) && (v[left] < v[left+1]))
    {
      left = left + 1;
    }

    if (v[i] > v[left]) return;
    
    swap(v, i, left);
    i = left;
  }
}


void build_heap(vector<int> &v)
{
  int i = v.size()/2 - 1;
  while (i >= 0)
  {
    heapify(v, i, v.size());
    i = i - 1;
  }
}


void heapsort(vector<int> &v)
{
  build_heap(v);

  int n = v.size() - 1;
  while (n > 0)
  {
    swap(v, 0, n);
    n = n - 1;
    heapify(v, 0, n);
  }
}

void print(const vector<int> &v)
{
  for (int i=0; i<v.size(); ++i)
  {
    cout << v[i] << " ";
  }
  cout << endl;
}
